Fix active nav menu markup

gadenbuie opened the pull request:
Fixes #411

This is a minimal fix for the behavior described in #411. The root issue is that BS5 requires a `.tab-pane.active` to have a corresponding `.active` toggle item. I'm being vague because the active class can appear as either `li.nav-item.active` or `.nav-link.active`. When the active class is imbalanced in the markup in BS5, the `.active` class isn't removed from the imbalanced `.tab-pane` until after it's activated.

https://user-images.githubusercontent.com/5420529/220937196-748eb254-4f90-44f0-93b3-46c2736bf4d8.mp4

The clip above uses the markup created by flexdashboard during initialization just [before the Bootstrap tab plugin is invoked](https://github.com/rstudio/flexdashboard/blob/bf8cdc55e15e3a7cfc53ed2d186a9303d271dc48/inst/www/flex_dashboard/flexdashboard.js#L135-L136). You can observe the tab behavior in Bootstrap versions 3-5 using this Rmd:

<details><summary><code>bootstrap-tabs-init.Rmd</code></summary>

````markdown
---
title: BS Nav Test
output:
  html_document:
    theme:
      version: 5
    self_contained: false
---

> Bootstrap version <span id="bs-version">x.y.z</span>

<span id="state">⏱️ before initializing tabs</span>

## No active classes anywhere

```{=html}
<div id="no-active">
  <div class="navbar bg-light navbar-default">
    <ul class="nav navbar-nav navbar-left">
      <li class="nav-item">
        <a class="nav-link" href="#no-active-one" data-toggle="tab">One</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="#no-active-two" data-toggle="tab">Two</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="#no-active-three" data-toggle="tab">Three</a>
      </li>
    </ul>
  </div>

  <div class="tab-content" style="height: 100%;">
    <div class="dashboard-page-wrapper tab-pane" id="no-active-one" style="height: 100%;">one content</div>
    <div class="dashboard-page-wrapper tab-pane" id="no-active-two" style="height: 100%;">two content</div>
    <div class="dashboard-page-wrapper tab-pane" id="no-active-three" style="height: 100%;">three content</div>
  </div>
</div>
```

## Nav doesn't have active class

```{=html}
<div id="nav-no-class">
  <div class="navbar bg-light navbar-default">
    <ul class="nav navbar-nav navbar-left">
      <li class="nav-item">
        <a class="nav-link" href="#nav-no-class-one" data-toggle="tab">One</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="#nav-no-class-two" data-toggle="tab">Two</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="#nav-no-class-three" data-toggle="tab">Three</a>
      </li>
    </ul>
  </div>

  <div class="tab-content" style="height: 100%;">
    <div class="dashboard-page-wrapper tab-pane active" id="nav-no-class-one" style="height: 100%;">one content</div>
    <div class="dashboard-page-wrapper tab-pane" id="nav-no-class-two" style="height: 100%;">two content</div>
    <div class="dashboard-page-wrapper tab-pane" id="nav-no-class-three" style="height: 100%;">three content</div>
  </div>
</div>
```

## Nav has active class on li

```{=html}
<div id="nav-class-li">
  <div class="navbar bg-light navbar-default">
    <ul class="nav navbar-nav navbar-left">
      <li class="nav-item active">
        <a class="nav-link" href="#nav-class-li-one" data-toggle="tab">One</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="#nav-class-li-two" data-toggle="tab">Two</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="#nav-class-li-three" data-toggle="tab">Three</a>
      </li>
    </ul>
  </div>

  <div class="tab-content" style="height: 100%;">
    <div class="dashboard-page-wrapper tab-pane active" id="nav-class-li-one" style="height: 100%;">one content</div>
    <div class="dashboard-page-wrapper tab-pane" id="nav-class-li-two" style="height: 100%;">two content</div>
    <div class="dashboard-page-wrapper tab-pane" id="nav-class-li-three" style="height: 100%;">three content</div>
  </div>
</div>
```

## Nav has active class on li > a

```{=html}
<div id="nav-class-li-a">
  <div class="navbar bg-light navbar-default">
    <ul class="nav navbar-nav navbar-left">
      <li class="nav-item">
        <a class="nav-link active" href="#nav-class-li-a-one" data-toggle="tab">One</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="#nav-class-li-a-two" data-toggle="tab">Two</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="#nav-class-li-a-three" data-toggle="tab">Three</a>
      </li>
    </ul>
  </div>

  <div class="tab-content" style="height: 100%;">
    <div class="dashboard-page-wrapper tab-pane active" id="nav-class-li-a-one" style="height: 100%;">one content</div>
    <div class="dashboard-page-wrapper tab-pane" id="nav-class-li-a-two" style="height: 100%;">two content</div>
    <div class="dashboard-page-wrapper tab-pane" id="nav-class-li-a-three" style="height: 100%;">three content</div>
  </div>
</div>
```

```{=html}
<button id="activate" class="btn btn-primary" style="position: absolute; bottom: 1em; right: 1em;">activate</button>
```

```{js echo=FALSE}
$(function() {
  const bs_v = $.fn.tab.Constructor.VERSION;
  console.log(`Bootstrap version: ${bs_v}`);
  document.getElementById('bs-version').innerText = bs_v;
})

$('#activate').on('click', function() {
  ['no-active', 'nav-no-class', 'nav-class-li', 'nav-class-li-a']
    .forEach(id => {
      console.log(
        `activating #${id}, active elements:`,
        document.querySelectorAll(`#${id} .active`)
      )
      $(`[href="#${id}-two"]`).tab('show')
    })

  $('#state').text('✅ tabs ready')
})
```

````

</details>

Testing across all the three major versions of Bootstrap reveals that adding the `.active` class to the nav list item maintains consistent appearance and behavior across all versions, even if the tab plugin were to fail for some reason.

| BS | Before | After |
|:--:|:--:|:--:|
| 3 | ![bs3-before](https://user-images.githubusercontent.com/5420529/220939006-03eaec58-6f7f-41c7-b25c-f33eaa0bac0b.png) | ![bs3-after](https://user-images.githubusercontent.com/5420529/220939030-65be7bfa-07f9-4fdd-a066-fe0e07087674.png) |
| 4 | ![bs4-before](https://user-images.githubusercontent.com/5420529/220939085-a566e437-8822-494b-889e-c1cfba587698.png) | ![bs4-after](https://user-images.githubusercontent.com/5420529/220939101-9d0d3519-5733-422b-93c1-8a6f9dd22470.png) |
| 5 | ![bs5-before](https://user-images.githubusercontent.com/5420529/220939179-adfe3590-a580-489d-a321-b1616ca5bdc7.png) | ![bs5-after](https://user-images.githubusercontent.com/5420529/220939194-4334a9e2-72c8-40be-a4f8-f89ae4492da4.png) |

cpsievert commented on inst/www/flex_dashboard/flexdashboard.js:
```diff
@@ -299,6 +297,12 @@ var FlexDashboard = (function () {
     }
 
+    // mark active tab and corresponding nav menu item
+    if (active) {
+      wrapper.addClass('active');
+      li.addClass('active');
+    }
```
Apologies for just realizing this, but this feels slightly cleaner/better?

```suggestion
    if (active)
      $(a).tab("show");
```

gadenbuie commented:
Now that we're calling `.tab('show')` in the navbar construction, the only code path that I can find where the call to `FlexDashboardUtils.manageActiveNavbarMenu()` is in the `.setLocation()` method of the utils.

https://github.com/rstudio/flexdashboard/blob/01b0bba637276bb9bbbf666874b00089f12daa5e/inst/www/flex_dashboard/flexdashboard.js#L1318-L1328

In every other case where `.manageActiveNavbarMenu()` is being called, we first call `.tab('show')` before the navbar is updated. My read on `.setLocation()` is that it's a way to change the page state that's driven by the content markup. In most cases, when `.tab('show')` is invoked first, the markup is already in the right place and `.setLocation()` is redundant. In the case where the active page is indicated via `.dashboard-page-wrapper.active` in a way that isn't reflected in the page URL, calling `.setLocation()` directly can update the page URL and change the active tab.

I'm... not sure why you'd want to do that, though. And looking through the change history it seems like maybe this is an artifact of how the navbar was managed initially? 

@cpsievert Anyway, I'm comfortable with this minimal fix for now and we can revisit the above notes sooner or later.

cpsievert commented:
@gadenbuie thanks for doing more research and the helpful write-up. At this point, I think I'd be on board for removing it. If we do remove it though, lets maybe do as a part of this PR so that the history/discussion isn't spread across multiple commits/threads?

gadenbuie commented:
> At this point, I think I'd be on board for removing it.

@cpsievert sounds good to me! I've removed the method and added notes in a couple of places where I learned things about the implementation that would be nice to remember.

I also discovered #420 while testing this, but I think that's likely better handled in a separate PR or in an upstream package.

cpsievert commented on inst/www/flex_dashboard/flexdashboard.js:
```diff
@@ -1236,16 +1239,18 @@ var FlexDashboard = (function () {
       var baseUrl = FlexDashboardUtils.urlWithoutHash(window.location.href);
       var hash = FlexDashboardUtils.urlHash($(this).attr('href'));
       var href = baseUrl + hash;
+      // the BS tab plugin handles changing the page and managing the navbar
+      // but flexdashboard needs to update the location and history
```
I find this to be a bit redundant given the comment above (but I don't feel strongly about it, so feel free to keep it)

```suggestion
```

cpsievert reviewed (approved):
Thanks, feel free to merge after updating the news!

Also, it's starting to feel like we should consider putting tests (either here or in `{shinycoreci}`) for these changes as well as for https://github.com/rstudio/flexdashboard/pull/421, but we can discuss that after merging.
